aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,272 @@
+
+Route templates must be unambiguous, or they will be ignored. For example, consider these two routes:
+
+1. ``{blog}/{year}/{month=1}``
+2. ``{blog}/{year}``
+
+The second route will never be used, since anything it would match will be matched by the first route (because it has a default value for month). If you wanted to construct a separate route that would for instance include all months or ask the user to select a month, you could change the route to include constant values, like this:
+
+1. ``monthly/{blog}/{year}/{month=1}``
+2. ``annual/{blog}/{year}``
+
+.. note:: There is a special case route for filenames, such that you can define a route value like ``files/{filename}.{ext?}``. When both ``filename`` and ``ext`` exist, both values will be populated. However, if only ``filename`` exists in the URL, the trailing period ``.`` is also optional. Thus, these would both match: ``/files/foo.txt`` and ``/files/foo``.
+
+.. tip:: Enable :doc:`logging` to see how the built in routing implementations, such as ``TemplateRoute``, match requests.
+
+Route Builder Extensions
+------------------------
+Several `extension methods on RouteBuilder <https://github.com/aspnet/Routing/blob/1.0.0-rc1/src/Microsoft.AspNet.Routing/RouteBuilderExtensions.cs>`_ are available for convenience. The most common of these is ``MapRoute``, which allows the specification of a route given a name and template, and optionally default values, constraints, and/or data tokens. When using these extensions, you must have specified the ``DefaultHandler`` and ``ServiceProvider`` properties of the ``RouteBuilder`` instance to which you're adding the route. These extensions all make use of the ``TemplateRoute`` type to add new route templates to whichever ``IRouter`` is configured to as the ``DefaultHandler``.
+
+.. note:: ``MapRoute`` doesn't take an ``IRouter`` parameter - it only adds routes to whatever the ``DefaultHandler`` is. This means that if the default is itself a route with named route value placeholders, only these placeholders can be used by routes defined using ``MapRoute``.
+
+If you want ``MapRoute`` to work for any route template, you'll need to ensure the ``DefaultHandler`` is set to use an ``IRouter`` instance that can accept any value. The following ``DelegateRouter`` accepts any values and displays them:
+
+.. literalinclude:: routing/sample/RoutingSample/DelegateRouter.cs
+  :linenos:
+  :language: c#
+
+You configure this ``IRouter`` implementation as the ``DefaultHandler`` for a ``RouteBuilder`` in the ``Configure`` method in ``Startup``:
+
+.. literalinclude:: routing/sample/RoutingSample/Startup.cs
```

Comment:
This snippet is a bit noisy. Ideally I think it would just show the route builder being created, setting the default handler and service provider, calling MapRoute and then calling UseRouter.